.github/workflows/zfs-build.yaml

```
@@ -0,0 +1,218 @@
+    - name: Tests
+      run: |
+        sudo mkdir -p $TEST_DIR
+        # run for 10 minutes or at most 2 iterations for a maximum runner time of 20 minutes.
+        sudo /usr/share/zfs/zloop.sh -t 600 -I 2 -l -m1 -- -T 120 -P 60
+    - name: Prepare artifacts
+      if: failure()
+      run: |
+        sudo chmod +r -R $TEST_DIR/
+    - uses: actions/upload-artifact@v3
+      if: failure()
+      with:
+        name: Logs-${{ matrix.os }}
+        path: |
+          /var/tmp/zloop/*/
+          !/var/tmp/zloop/*/vdev/
+        if-no-files-found: ignore
+    - uses: actions/upload-artifact@v3
+      if: failure()
+      with:
+        name: Pool files-${{ matrix.os }}
+        path: |
+          /var/tmp/zloop/*/vdev/
+        if-no-files-found: ignore
+  tests-sanity:
+    name: Sanity tests
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [18.04, 20.04 ]
+    runs-on: ubuntu-${{ matrix.os }}
+    needs: build
+    steps:
+    - uses: actions/download-artifact@v3
+      with:
+        name: ubuntu-${{ matrix.os }}
+    - name: Install dependencies
+      run: |
+        tar xzvf ubuntu-${{ matrix.os }}.tgz
+        sudo apt-get update
+        xargs --arg-file=.github/workflows/test-dependencies.txt sudo apt-get install -qq
+        sudo apt-get clean
+        sudo dpkg -i *.deb
+        sudo sed -i.bak 's/updates/extra updates/' /etc/depmod.d/ubuntu.conf
+        sudo depmod
+        sudo modprobe zfs
+    - name: Reclaim disk space
+      run: |
+        .github/workflows/scripts/reclaim_disk_space.sh
+    - name: Clear the kernel ring buffer
+      run: |
+        sudo dmesg -c >/var/tmp/dmesg-prerun
+    - name: Report CPU information
+      run: |
+        lscpu
+        cat /proc/spl/kstat/zfs/chksum_bench
+    - name: Report disk space
+      run: |
+        df -h /
+    - name: Running sanity tests
+      run: |
+        set -o pipefail
+        /usr/share/zfs/zfs-tests.sh -vKR -s 3G -r sanity | scripts/zfs-tests-color.sh
+      shell: bash
+      timeout-minutes: 330
+    - name: Prepare artifacts
+      if: failure()
+      run: |
+        RESULTS_PATH=$(readlink -f /var/tmp/test_results/current)
+        sudo dmesg > $RESULTS_PATH/dmesg
+        sudo cp /var/log/syslog /var/tmp/dmesg-prerun $RESULTS_PATH/
+        sudo chmod +r $RESULTS_PATH/*
+        # Replace ':' in dir names, actions/upload-artifact doesn't support it
+        for f in $(find /var/tmp/test_results -name '*:*'); do mv "$f" "${f//:/__}"; done
+    - uses: actions/upload-artifact@v3
+      if: failure()
+      with:
+        name: Test logs Ubuntu-${{ matrix.os }}
+        path: |
+          /var/tmp/test_results/*
+          !/var/tmp/test_results/current
+        if-no-files-found: ignore
+  tests-functional:
+    name: Functional tests
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [18.04, 20.04 ]
+        tests: [ part1, part2, part3, part4, part5 ]
+    runs-on: ubuntu-${{ matrix.os }}
+    needs: tests-sanity
```

`Sanity tests` is a subset from `Functional tests` just to get basic results faster. So I propose to leave them run in parallel
```suggestion
    needs: build
```

---

The two sanity tests are run in parallel after building the modules and need about 25m.
The idea is, when the sanity tests are failing, we have no need to start the functional tests et all.

---

> we have no need to start the functional tests et all.

But we can get more failed tests without need to wait for sanity tests' green run next time. Otherwise there will be a one more tests fixing and CI waiting cycle guaranteed for developer. More time to spend - less motivation.

---

As far as i know we don't have to save any runners' resources of github actions, it's free and we may not want to trade free cpu time for a contributor expensive time, this is my concern.

---

Good point ... I should adjust the needs flags.

---

Aaand if we are on a road of splitting tests per parallel jobs, we can exclude `sanity tests` group from full runners after all, but it'll be harder than just change running tags of tests (thought for future)